### PR TITLE
Add `order` parameter to `backend.calculate_norm`

### PR DESCRIPTION
**Pull request by renatomello**
Checklist:
- [x] Reviewers confirm new code works as expected.
- [x] Tests are passing.
- [x] Coverage does not decrease.
- [x] Documentation is updated.

### CODE REVIEW

File: src/qibo/backends/tensorflow.py [hunk 1]
```diff
@@ -306,6 +306,20 @@ def entanglement_entropy(self, rho):
         entropy = self.np.sum(masked_eigvals * spectrum) / self.np.log(2.0)
         return entropy, spectrum
 
+    def calculate_norm(self, state, order=None):
+        state = self.cast(state)
```
Comment (MatteoRobbiati) [inline src/qibo/backends/tensorflow.py:310]: Maybe here you can directly use [`tf.norm`](https://www.tensorflow.org/api_docs/python/tf/norm), which seems to be quite similar to `np.linalg.norm`.

File: src/qibo/backends/numpy.py [hunk 2]
```diff
@@ -673,13 +673,13 @@ def entanglement_entropy(self, rho):
...
-    def calculate_norm(self, state):
+    def calculate_norm(self, state, order=None):
         state = self.cast(state)
-        return self.np.sqrt(self.np.sum(self.np.abs(state) ** 2))
+        return self.np.linalg.norm(state, ord=order)
 
-    def calculate_norm_density_matrix(self, state):
+    def calculate_norm_density_matrix(self, state, order="nuc"):
         state = self.cast(state)
-        return self.np.trace(state)
```
Comment (MatteoRobbiati) [inline src/qibo/backends/numpy.py:682]: Is one of the order options corresponding to the trace?

File: src/qibo/backends/tensorflow.py (hunk 1 above)
Comment (renatomello) [reply src/qibo/backends/tensorflow.py:310]: I tried that too and implemented it but then moved back to `numpy` because `tf.norm` not only uses orders with different names, but also doesn't implement all the orders that `numpy` has. So, I kept using the `numpy` norm in order to unify the `order` parameter with all the other backends.

File: src/qibo/backends/numpy.py (hunk 2 above)
Comment (renatomello) [reply src/qibo/backends/numpy.py:682]: Nuclear norm (also known as trace norm)

File: src/qibo/backends/tensorflow.py (hunk 1 above)
Comment (MatteoRobbiati) [reply src/qibo/backends/tensorflow.py:310]: > I tried that too and implemented it but then moved back to `numpy` because `tf.norm` not only uses orders with different names, but also doesn't implement all the orders that `numpy` has. So, I kept using the `numpy` norm in order to unify the `order` parameter with all the other backends.

Okay, I understand. The only problem I see is that this is not compatible with TensorFlow gradient descent in simulation. I Think it will lose track of the state using an unknown  (by TensorFlow) numpy function.

Comment (renatomello) [reply src/qibo/backends/tensorflow.py:310]: > > I tried that too and implemented it but then moved back to `numpy` because `tf.norm` not only uses orders with different names, but also doesn't implement all the orders that `numpy` has. So, I kept using the `numpy` norm in order to unify the `order` parameter with all the other backends.
> 
> Okay, I understand. The only problem I see is that this is not compatible with TensorFlow gradient descent in simulation. I Think it will lose track of the state using an unknown (by TensorFlow) numpy function.

you mean if `backend.calculate_norm` is used as a callback?

Comment (MatteoRobbiati) [reply src/qibo/backends/tensorflow.py:310]: Yes, in general if you want to calculate gradients of a process which involves non-tensorflow objects, it can lose the track of the flow and return `None`.

Comment (renatomello) [reply src/qibo/backends/tensorflow.py:310]: Do we have any test covering that right now? If so, shouldn't it be broken? If not, we need to have one then.

Comment (MatteoRobbiati) [reply src/qibo/backends/tensorflow.py:310]: I don't think we have tests for every scenario right now (and I don't think it is possible).

My only concern is that if we built the TensorFlow backend to do automatic differentiation, it makes sense to keep the total compatibility with TF, because I suppose if one decide to use `tensorflow` as backend is because some gradient descent has to be computed. 

With this in mind, maybe I would like to customize the method if `tensorflow` is set, in order to remove the `"order"` option and manually compute only one norm if TF is used.

Maybe @stavros11 has clearer picture of this.

Comment (renatomello) [reply src/qibo/backends/tensorflow.py:310]: We could also just use `tf.norm` and let the user know that not all orders in the other backends are compatible with the `tensorflow` backend. It is non-uniform but it works.

Comment (MatteoRobbiati) [reply src/qibo/backends/tensorflow.py:310]: > We could also just use `tf.norm` and let the user know that not all orders in the other backends are compatible with the `tensorflow` backend. It is non-uniform but it works.

Yes, it is also a good solution.

Comment (renatomello) [reply src/qibo/backends/tensorflow.py:310]: > Yes, it is also a good solution.

Done

File: src/qibo/backends/abstract.py [hunk 3]
```diff
@@ -326,13 +326,29 @@ def entanglement_entropy(self, rho):  # pragma: no cover
...
+        """
         raise_error(NotImplementedError)
 
     @abc.abstractmethod
-    def calculate_norm_density_matrix(self, state):  # pragma: no cover
-        """Calculate norm (trace) of a density matrix."""
+    def calculate_norm_density_matrix(self, state, order="nuc"):  # pragma: no cover
+        """Calculate norm (trace) of a density matrix.
+
+        If ``order="nuc"``, it returns the nuclear norm os ``state``,
```
Comment (MatteoRobbiati) [inline src/qibo/backends/abstract.py:344]:
```suggestion
        If ``order="nuc"``, it returns the nuclear norm of ``state``,
```

File: src/qibo/backends/tensorflow.py [hunk 4]
```diff
@@ -306,6 +306,16 @@ def entanglement_entropy(self, rho):
...
         return entropy, spectrum
 
+    def calculate_norm(self, state, order=2):
+        state = self.cast(state)
+        return self.tf.norm(state, ord=order)
+
+    def calculate_norm_density_matrix(self, state, order="nuc"):
+        state = self.cast(state)
+        if order == "nuc":
+            return self.np.trace(state)
```
Comment (stavros11) [inline src/qibo/backends/tensorflow.py:316]: Regarding the discussion on using tensorflow primitives, indeed, if someone uses gradient descent over a numpy operation it will break the gradient calculation. However, note that in this case `self.np` is not `numpy` but
```py
import tensorflow.experimental.numpy as tnp
```
(see the `__init__`). Back at the time we were writing this backend this was an easy way to port numpy operations to tensorflow, but I am not sure what is the status now. I am also not sure how it will behave in terms of the gradient.

File: src/qibo/backends/abstract.py [hunk 5]
```diff
@@ -326,13 +326,29 @@ def entanglement_entropy(self, rho):  # pragma: no cover
         raise_error(NotImplementedError)
 
     @abc.abstractmethod
-    def calculate_norm(self, state):  # pragma: no cover
-        """Calculate norm of a state vector."""
+    def calculate_norm(self, state, order=2):  # pragma: no cover
+        """Calculate norm of a state vector.
```
Comment (MatteoRobbiati) [inline src/qibo/backends/abstract.py:330]:
```suggestion
        """Calculate norm of a state vector. The 2-norm is computed as default option. 
```

**Review by MatteoRobbiati — APPROVED** [review]
Thanks! It looks good to me. Just two little suggestions follow.

File: src/qibo/backends/tensorflow.py (hunk 4 above)
Comment (MatteoRobbiati) [reply src/qibo/backends/tensorflow.py:316]: Indeed! 
This works well.
